SYCL/ESIMD/api/functional/value.hpp

```diff
@@ -135,41 +156,76 @@ template <typename DataT, int NumElems> std::vector<DataT> generate_ref_data() {
-      }
-    }
+    ref_data = details::construct_ref_data<DataT, NumElems>(
+        0.25, {-inf, nan, min, max, -0.0, 0.1}, {-0.1, +0.0});
   } else if constexpr (std::is_signed_v<DataT>) {
-    ref_data.reserve((NumElems > 1) ? NumElems : 5);
-
-    ref_data.insert(ref_data.end(), {min, min_half, max, max_half, 0});
-    if constexpr (NumElems != 1) {
-      ref_data.insert(ref_data.end(), {min_plus_one, max_minus_one, -1});
-      for (size_t i = ref_data.size(); i < NumElems; ++i) {
-        ref_data.push_back(i);
-      }
-    }
+    ref_data = details::construct_ref_data<DataT, NumElems>(
+        0, {min, min_half, max, max_half, 0},
+        {min_plus_one, max_minus_one, -1});
   } else {
-    // Unsigned integral type
-    ref_data.reserve((NumElems > 1) ? NumElems : 3);
-
-    ref_data.insert(ref_data.end(), {max, max_half, 0});
-    if constexpr (NumElems != 1) {
-      ref_data.insert(ref_data.end(), {max_minus_one});
-      for (size_t i = ref_data.size(); i < NumElems; ++i) {
-        ref_data.push_back(i);
-      }
-    }
+    ref_data = details::construct_ref_data<DataT, NumElems>(
+        0, {max, max_half, 0}, {max_minus_one});
+  }
+  return ref_data;
+}
+
+// Provides std::vector with the reference data according to the obtained data
+// types and number of elements.
+template <typename SrcT, typename DstT, int NumElems>
+std::vector<SrcT> generate_ref_conv_data() {
+  static_assert(std::is_integral_v<SrcT> ||
+                    type_traits::is_sycl_floating_point_v<SrcT>,
+                "Invalid the first data type provided to the "
+                "generate_ref_conv_data function as source type.");
+  static_assert(std::is_integral_v<DstT> ||
+                    type_traits::is_sycl_floating_point_v<DstT>,
+                "Invalid the second data type provided to the "
+                "generate_ref_conv_data function as destination type.");
+
+  static const SrcT min =
+      std::max(value<SrcT>::lowest(), static_cast<SrcT>(value<DstT>::lowest()));
+  static const SrcT min_half = min / 2;
+  static const SrcT max =
+      std::min(value<SrcT>::max(), static_cast<SrcT>(value<DstT>::max()));
+  static const SrcT max_half = max / 2;
+  static const SrcT nan = value<SrcT>::nan();
+  static const SrcT inf = value<SrcT>::inf();
+  static const SrcT denorm = std::max(
+      value<SrcT>::denorm_min(), static_cast<SrcT>(value<DstT>::denorm_min()));
+  static const SrcT positive = static_cast<SrcT>(126.75);
+
+  std::vector<SrcT> ref_data{};
+
+  if constexpr (type_traits::is_sycl_floating_point_v<SrcT> &&
+                type_traits::is_sycl_floating_point_v<DstT>) {
+    ref_data = details::construct_ref_data<SrcT, NumElems>(
+        0.25, {min, max, -0.0, +0.0, 0.1, denorm, nan, -inf});
+  } else if constexpr (type_traits::is_sycl_floating_point_v<SrcT> &&
+                       std::is_unsigned_v<DstT>) {
+    ref_data = details::construct_ref_data<SrcT, NumElems>(
+        0.75, {-0.0, max, max_half, -max_half});
```

Comment:
So we basically have `0.75`, `0.25` and `0` for the fractional part of reference data tail. Given the use cases we could talk about `0.75` and `0.25` values. The only reason we want `0.75` is to verify that fractional part doesn't affect the conversion to integral type at all, that there is no some "upper bound" rounding. The only reason we have `0.25` for is to have some fractional part, to make sure it preserves.

So what if we just use `0.75` everywhere, with no additional parameter passed to `details::construct_ref_data`? Are there any caveats I don't see?

Reply:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/696/commits/c9bc321af2b1bf1b9c70b323c48041dd1e79a85e#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514L162-L171)